GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs

```diff
@@ -202,6 +202,12 @@ protected void OnWorkingDirectoryFileOrFolderDeleteNotification(string relativeP
                         this.FileSystemCallbacks.OnFolderDeleted(relativePath);
                     }
                 }
+                else if (!isPreDelete)

```

Comment:
A few suggestions/considerations, let me know what you think:


1. It could be a bit misleading/confusing to call a method named `OnTombstoneFolderCreated` here.  When `OnWorkingDirectoryFileOrFolderDeleteNotification` is called on Windows there's no guarantee that a tombstone has been created, it might have simply been a full folder that was deleted.

    One way to address this would be to add another parameter (something like `triggeredByGit`) to `OnFolderDeleted`, call `OnFolderDeleted` instead, and then push the decision of how to handle deletes by git into `OnFolderDeleted`.

2. Whether it be here or in `FileSystemCallbacks.OnFolderDeleted`, should we only record the folder delete when the platform supports tombstones?  If Linux ends up supporting post-delete notifications and doesn't use tombstones there would be no need to record the delete in the placeholder database.  Perhaps rather than driving this decision based on `isPreDelete` we should check `GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories` (or maybe add a new `GVFSPlatform.Instance.KernelDriver.SupportsTombstones`).

3. Related to 1, because there's no guarantee that this path is a tombstone, should we record something other than `TOMBSTONE FOLDER` in the placeholder database (e.g. `PSBL_TOMBSTONE FOLDER`)?  An alternative would be to have a separate database of deleted folders that we check against when updating folder placeholders, but that seemed potentially more complicated (and slower).

Reply:
>push the decision of how to handle deletes by git into OnFolderDeleted

I don't think we would have enough information there either unless we try to read the directory attributes to find the tombstone attribute.  So I don't see a difference in where the decision is made.  
>only record the folder delete when the platform supports tombstones

What do you think about having the `FileSystemVirtualizer` have a method that is virtual and doesn't do anything something like `OnPossibleTombstoneFolder` and override that in the `WindowsFileSystemVirtualizer` then we wouldn't need anything extra in the Platform.

I don't really want another database to keep this data since it is about a placeholder and the placeholder database is where we keep that information.  Adding more information into the `SHA` field is fine.  We have 40 characters. :)

Reply:
> What do you think about having the FileSystemVirtualizer have a method that is virtual and doesn't do anything something like OnPossibleTombstoneFolder and override that in the WindowsFileSystemVirtualizer then we wouldn't need anything extra in the Platform.

I can get on board with that if the MacFileSystemVirtualizer has a no-op implementation with a comment explaining why it's not doing anything.  However, it still feels a bit wrong to me to call `OnPossibleTombstoneFolder` on a platform that doesn't support it, and so let me throw another alternative out there and let me know what you think.

We could remove the shared `OnWorkingDirectoryFileOrFolderDeleteNotification` method and move the "isPreDelete" code into `MacFileSystemVirtualizer` and the "post delete + tombstone" code to `WindowsFileSystemVirtualizer`. 

`OnWorkingDirectoryFileOrFolderDeleteNotification` was first added because the delete code on the two platforms was so similar, but perhaps it's diverging enough now it shouldn't be shared anymore?

Reply:
I will give that a try and see how it looks.  They are different in that they each will call different `FileSystemCallbacks` but they share the logic for when to call them but this change is kind of breaking that.

Reply:
@wilbaker I reworked this to try and keep the logic and have the platform specific code handle the difference in what `FileSystemCallbacks` was called.  Let me know what you think and if there are any tweaks to the method names I should make.